src/main.rs

```diff
@@ -21,15 +21,17 @@ pub mod input_editor;
 pub mod peg;
 pub mod expansion;
 
-pub type Variables = BTreeMap<String, String>;
+pub struct Mode {
+    value: bool,
+}
 
 /// This struct will contain all of the data structures related to this
 /// instance of the shell.
 pub struct Shell {
-    pub variables: Variables,
-    pub modes: Vec<Mode>,
-    pub directory_stack: DirectoryStack,
-    pub history: VecDeque<String>,
+    variables: BTreeMap<String, String>,
```

Comment:
Actually, it was you that made the suggestion to BTreeMap, so that vars should also be automatically sorted upon insert. :stuck_out_tongue_closed_eyes: